### PR TITLE
feat: move garbage collection into a separate thread

### LINKED ISSUE #10971: Garbage collection may take a long time
A validator reported that their node usually miss blocks and/or chunks at the beginning of an epoch. At some point in the middle of the epoch after gc finishes, the number of missed blocks and chunks starts to go down. As can be see [in grafana](https://nearinc.grafana.net/d/zzGDUdcnk/debug-metrics?orgId=1&var-chain_id=mainnet&var-role=rpc&var-node_type=rpc&var-node_id=All&var-shard_id=All&viewPanel=61&from=1712446819503&to=1712533219503), GC can take more than 0.3s, which could make the difference between missing a block/chunk vs. producing them on time.

A long term solution is to move GC to a different thread, as attempted in https://github.com/near/nearcore/pull/3090. However, that could require some nontrivial work and a short term mitigation could be that we don't run garbage collection when block processing already takes a relatively long time.

### PR DESCRIPTION
**Pull request by bowenwang1996**
Move garbage collection into a separate actor to prevent it from blocking synchronously inside client actor. Fixes #10971. For testing, `garbage_collection_intense.py` sends a lot of insertion and deletion transactions on the same key and make sure that nodes do not crash after gc is done. The change is also run on a mainnet node and so far it works fine.

TODO: run all nayduck tests

### CODE REVIEW

**Comment by bowenwang1996** [comment]
@wacban @posvyatokum what is the best way to test this for archival nodes with split storage?

**Comment by posvyatokum** [comment]
Off the top of my head

To make sure that cold loop is working properly:
- Run canary and check that cold head is increasing, and it is smaller than tail.

To make sure that we didn't accidentally deleted important data before backup without impacting cold loop: 
- Replaying history
- Checking column contents against prod split storage nodes
- Traversing Trie

File: chain/client/src/gc_actor.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+use actix::{Actor, Addr, Arbiter, ArbiterHandle, AsyncContext, Context};
+use near_chain::{types::RuntimeAdapter, ChainStore, ChainStoreAccess};
+use near_chain_configs::{ClientConfig, GCConfig};
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::types::BlockHeight;
+use near_store::{metadata::DbKind, Store};
+use std::sync::Arc;
+use tracing::warn;
+
+pub struct GCActor {
```
Comment (wacban) [inline chain/client/src/gc_actor.rs:11]: nit: Add a comment please. 

random remark: I think it makes sense to have gc in a dedicated actor instead of adding it to e.g. to the SyncJobsActor to make sure it's not blocked by any lengthy job that may be sitting there (loading trie, resharding..).

File: chain/client/src/gc_actor.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            runtime_adapter,
+            gc_config,
+            epoch_manager,
+            is_archive,
+        }
+    }
+
+    fn gc(&mut self, ctx: &mut Context<GCActor>) {
+        let timer = metrics::GC_TIME.start_timer();
+        if !self.is_archive {
```
Comment (wacban) [inline chain/client/src/gc_actor.rs:39]: nit: I think this matches the deleted Chain::clear_data method, can you move the comments from there here?

File: chain/client/src/gc_actor.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                        self.runtime_adapter.clone(),
+                        self.epoch_manager.clone(),
+                    ) {
+                        warn!(target: "client", "Error in gc: {}", e);
+                    }
+                }
+                Err(e) => {
+                    warn!(target: "client", "Error in gc: {}", e);
+                }
+                _ => {}
```
Comment (wacban) [inline chain/client/src/gc_actor.rs:64]: I believe for other DbKind types you should call the `clear_archive_data` method.

Comment (wacban) [reply chain/client/src/gc_actor.rs:64]: cc @posvyatokum who's an expert and a recent editor of DbKinds :)

File: pytest/tests/loadtest/contract/Cargo.toml [hunk 4]
```diff
@@ -4,9 +4,6 @@ version = "0.1.0"
 authors = ["Near Inc <hello@near.org>"]
 edition = "2018"
 
-[lints]
-workspace = true
-
```
Comment (wacban) [inline pytest/tests/loadtest/contract/Cargo.toml:9]: I'm not familiar with lints, what does this do and why did you delete it?

File: nightly/pytest-sanity.txt [hunk 5]
```diff
@@ -85,6 +85,8 @@ pytest --timeout=120 sanity/garbage_collection.py
 pytest --timeout=120 sanity/garbage_collection.py --features nightly
 pytest --timeout=120 sanity/garbage_collection1.py
 pytest --timeout=120 sanity/garbage_collection1.py --features nightly
+pytest --timeout=120 sanity/garbage_collection_intense.py
+pytest --timeout=120 sanity/garbage_collection_intense.py --features nightly
```
Comment (wacban) [inline nightly/pytest-sanity.txt:89]: I think you forgot to add this file to the PR.

**Review by wacban — CHANGES_REQUESTED** [review]
looks good overall and it's a nice idea

Please see comments - I think we need to handle archival nodes a bit differently. 

In order to trigger nayduck you can checkout this branch and run `./scripts/nayduck.py` from nearcore.

**Comment by bowenwang1996** [comment]
@wacban @posvyatokum I have an idea: instead of testing whether it works for archival nodes with split storage, we don't change the logic for archival nodes, i.e, they still move data from hot storage to cold storage as part of the main thread. I think this is okay because right now the main goal of archival node is to store historical data and it does not need to be able to process new blocks very fast. This would save us some trouble in testing archival nodes and we can potentially do the change for archival node as well at a later time, if necessary. What do you think?

**Comment by wacban** [comment]
> @wacban @posvyatokum I have an idea: instead of testing whether it works for archival nodes with split storage, we don't change the logic for archival nodes, i.e, they still move data from hot storage to cold storage as part of the main thread. I think this is okay because right now the main goal of archival node is to store historical data and it does not need to be able to process new blocks very fast. This would save us some trouble in testing archival nodes and we can potentially do the change for archival node as well at a later time, if necessary. What do you think?

Personally I don't like it as it would mean we have two entirely different code paths and even threads for garbage collection. The CI is failing so it seems like we have at least some test coverage. I would suggest fixing those first and if coverage is missing for split storage we should add it. 

If you want to test it on a real node that shouldn't be too hard as well, it takes about 10min to spin it up. Then I suppose it would be sufficient to add and check some logs.

File: chain/client/src/client_actor.rs [hunk 6]
```diff
@@ -218,6 +227,21 @@ pub fn start_client(
     )
     .unwrap();
     let resharding_handle = client.chain.resharding_handle.clone();
+
+    let maybe_gc_arbiter = {
```
Comment (shreyan-gupta) [inline chain/client/src/client_actor.rs:231]: I don't think it's this straight forward to create a new actor. We need to handle creating mocks for tests as well.

File: pytest/tests/loadtest/contract/Cargo.toml (hunk 4 above)
Comment (bowenwang1996) [reply pytest/tests/loadtest/contract/Cargo.toml:9]: Doesn't compile otherwise

File: integration-tests/src/tests/client/process_blocks.rs [hunk 7]
```diff
@@ -1433,122 +1429,6 @@ fn test_archival_save_trie_changes() {
     }
 }
 
-fn test_archival_gc_common(
```
Comment (wacban) [inline integration-tests/src/tests/client/process_blocks.rs:1436]: Why delete those tests? Can you run the gc from the main test thread manually to make it work?

**Review by wacban — APPROVED** [review]
LGTM

File: integration-tests/src/tests/client/process_blocks.rs (hunk 7 above)
Comment (bowenwang1996) [reply integration-tests/src/tests/client/process_blocks.rs:1436]: I can, but I am not sure what that would achieve. Essentially it requires implementing a separate logic that manually invokes `clear_data` for archival nodes. However, if we want to test the underlying `clear_data` implementation, there are already integration tests that only constructs `Chain` which achieves that. If we want to test that archival node as a whole function as expected (doesn't delete data), then `test_garbage_collection_archival.py` achieves that. A test here feels like it is not testing any interfaces. If there is a bug in how `GCActor` runs, then it may very well happen that a test here passes but archival nodes do not work properly. I think what would help is to put `test_garbage_collection_archival.py` into CI. @andrei-near what is the best way to do that?

Comment (bowenwang1996) [reply integration-tests/src/tests/client/process_blocks.rs]: Fixed

**Comment by bowenwang1996** [comment]
I am going to merge this PR. For testing we have:
* [Nayduck run](https://nayduck.nearone.org/#/run/56)
* It has been running on a regular mainnet node for over a week without any issues
* It has been running on an archival mainnet node with split storage for 2 days and [cold store head](https://nearone.grafana.net/explore?schemaVersion=1&panes=%7B%22ob3%22:%7B%22datasource%22:%22grafanacloud-prom%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22near_cold_head_height%7Brole%3D%5C%22bowen-0%5C%22%7D%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22grafanacloud-prom%22%7D,%22editorMode%22:%22builder%22,%22legendFormat%22:%22__auto%22,%22useBackend%22:false,%22disableTextWrap%22:false,%22fullMetaSearch%22:false,%22includeNullMetadata%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-12h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1) increases as expected.
